### ISSUE TITLE
workspace protocol found in peerDeps

### ISSUE THREAD

**Issue by Mister-Hope** [lerna-lite/lerna-lite#869]
### Describe the bug

We choose to do version bump locally and let ci do the rest, while I am bumping the version with `lerna version`, and the ci released new version using `lerna publish --dist-tag next from-package --yes`. [GitHub action log here](https://github.com/vuepress/ecosystem/actions/runs/9250159557/job/25443201348)

I found the docs [here](https://github.com/lerna-lite/lerna-lite/tree/main/packages/version#--allow-peer-dependencies-update) before opening the issue, but it seems that I should not see workspace protocol anyway even without `--allow-peer-dependencies-update`

However, the following things are published as is:

```
  "peerDependencies": {
    "@vuepress/plugin-git": "workspace:~",
    "vuepress": "2.0.0-rc.12"
  },
```

See https://www.npmjs.com/package/@vuepress/plugin-append-date/v/2.0.0-rc.31?activeTab=code

![image](https://github.com/lerna-lite/lerna-lite/assets/33315834/600f1038-30ba-415d-b742-8ace2378d820)

### Used Package Manager

pnpm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/lerna-lite/lerna-lite/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [docs](https://github.com/lerna-lite/lerna-lite#readme).
- [X] Check that there isn't [already an issue](https://github.com/lerna-lite/lerna-lite/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/lerna-lite/lerna-lite/discussions).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by ghiscoding** [comment]
there's no point in setting `workspace:` protocol without enabling `--allow-peer-dependencies-update`, because if you're not bumping the peer dependencies then... use a static version. The best I could do is throw an error mentioning what I just said, apart from that I don't see the point.

**Comment by Mister-Hope** [comment]
Thanks for the quick fix,  I believe I catced a covid,  And I've been resting in the past few days.

About the ~workspace,  I was trying to specific any version with that is not less than this package in minor.  But yes, this might be breaking changes to users.

**Comment by ghiscoding** [comment]
oh well I hope you get well super quickly.

I'm not sure what you meant by the breaking change though, but again if you wish to use `workspace:` in the peer dependencies of your project then I suggest you use the `--allow-peer-dependencies-update` because I don't see any reasons to use them separately. At least now with the new version, it will tell you if you're not using them correctly

**Comment by Mister-Hope** [comment]
Hi， I am still confused, with the following questions:

1. What's the expected behavior of a `workspace:*` peer result after bumping with `lerna version`?
2. Does lerna support publishing it?

The problem is that now I am using `@vuepress/plugin-git: workspace:*`  in peer, but I got no issue with `lerna version`, and the published result is still containing `workspace:*`, which means:

- The PR change is invalid, at least to me
- lerna still fails to publish a pkg in the way that pnpm supports.

Furthermore, omit the `--allow-peer-dependencies-update` behavior arguing, the version bump and version release are separate steps, as long as the bumped result are still a valid one (meaningful one), then `lerna release` should try to publish just like I am doing `pnpm publish -r`(but only on needed packages), this behavior is definitely broken, no matter the previous `workspace:~` or current `workspace:*` in peer deps.

Suggesting opening this issue for futher discussion and fix.

**Comment by Mister-Hope** [comment]
Additional information: https://github.com/vuepress/ecosystem/commit/93df303baa6291f52e3523f136073e79d53274ea is released with just `lerna version` and `workspace:*` peers, no issues are thrown. The release result still have the as is `workspace:*` peers.

**Comment by ghiscoding** [comment]
Yes there should be an error when you omit `--allow-peer-dependencies-update`, it does show on my side when I tested it locally. I don't want to support peer dependencies without the flag and from what I can see in your project, you still did not enable the flag. I don't want to support workspace protocol without the flag enabled. So why are you looking for trouble? You should try with `--dry-run` mode before publishing any real versions.

https://github.com/lerna-lite/lerna-lite/blob/83622fbd755d3ce4f19a0ed213943258be48d0b7/packages/core/src/package.ts#L312-L319

**Comment by ghiscoding** [comment]
Actually I took another quick look and found that Lerna completely ignores peer dependencies unless you enable the flag, so... since you did not have the flag enabled then it makes sense that the `workspace:` was published because again the peer deps were completely ignored (not even read) without the flag. So yes you are right that the error is not being thrown, but again... why do you persist in not enabling the flag when it is the source of all your problem?

You can see on line 60 below that `peerDependencies` is completely skipped (not read) unless the `graphType` is `allPlusPeerDependencies`

https://github.com/lerna-lite/lerna-lite/blob/a944b522b66084888681c216cb303695d5eb46ab/packages/core/src/package-graph/package-graph.ts#L53-L62

and that `allPlusPeerDependencies` can only be true when `allowPeerDependenciesUpdate` flag is enabled

https://github.com/lerna-lite/lerna-lite/blob/a944b522b66084888681c216cb303695d5eb46ab/packages/core/src/command.ts#L323-L327

So in summary, yes you're right that the previous PR I've done doesn't actually throw... but I don't really how I could throw since the only time the peer deps are read is when the `allowPeerDependenciesUpdate` flag is enabled... so it's a chicken and egg problem (what came first). 

in short... just enable the `allowPeerDependenciesUpdate` flag please!!!

**Comment by Mister-Hope** [comment]
I read your reply 3 times, I understand now why no errors are thrown, but I am still confused about my last problem. Note, I already stop using `workspace:~` in rc stage and now everything is `workspace:*`

In short, I split the version bump `lerna vesion` and release packages `lerna publish` into 2 commands.

1. `lerna publish` does not have a such flag called `allowPeerDependenciesUpdate`, if you are trying to tell me that I must publish with `--graphType=allPlusPeerDependencies` flag, I will accept your advice to add a such flag in publish command. **But I insist this is a lerna bug as this is not what `pnpm publish` behaves**. Lerna should try to parse the command.

2. Bumping with or without this flag with `--dry-run` makes no difference to me, as long as lerna supports `workspace:*` and do not force converting this into a `workspace:version`, then I don't think it will affect the result when ci runs the publish command. And I ensure the docs says that this supported, and the `--dry-run` is not converting the `workspace:*` to something else. (show case in below)

To proof what I am talking, I create [this branch](https://github.com/vuepress/ecosystem/commits/lerna-test/). I see **no difference** adding this flag (yes, I tested already before questioning whether your fix is valid, it is not me that still insist to publish new versions without flag, just I can't see any difference with no changes and no errors)

So, I do not accept your announcement that with the flag then peerDeps are good. I belived that these versions will still be released leaving the `workspace:*` in peerDeps as is.

So at last I need to confirm:

- Are you suggesting I must add other things like `--graphType=allPlusPeerDependencies` to make it work
- Shall this be a bug while `pnpm publish` support this?

**Comment by Mister-Hope** [comment]
I read your reply 3 times, I understand now why no errors are thrown, but I am still confused about my last problem. Note, I already stop using `workspace:~` in rc stage and now everything is `workspace:*`

In short, I split the version bump `lerna vesion` and release packages `lerna publish` into 2 commands.

1. `lerna publish` does not have a such flag called `allowPeerDependenciesUpdate`, if you are trying to tell me that I must publish with `--graphType=allPlusPeerDependencies` flag, I will accept your advice to add a such flag in publish command. **But I insist this is a lerna bug as this is not what `pnpm publish` behaves**. Lerna should try to parse the command.

2. Bumping with or without this flag with `--dry-run` makes no difference to me, as long as lerna supports `workspace:*` and do not force converting this into a `workspace:version`, then I don't think it will affect the result when ci runs the publish command. And I ensure the docs says that this supported, and the `--dry-run` is not converting the `workspace:*` to something else. (show case in below)

To proof what I am talking, I create [this branch](https://github.com/vuepress/ecosystem/commits/lerna-test/). I see **no difference** adding this flag (yes, I tested already before questioning whether your fix is valid, it is not me that still insist to publish new versions without flag, just I can't see any difference with no changes and no errors)

So, I do not accept your announcement that with the flag then peerDeps are good. I believed that these versions will still be released leaving the `workspace:*` in peerDeps as is.

So at last I need to confirm:

- Are you suggesting I must add other things like `--graphType=allPlusPeerDependencies` to make it work
- Shall this be a bug while `pnpm publish` support this?

**Comment by ghiscoding** [comment]
you don't seem to understand that Lerna/Lerna-Lite **completely ignores peer dependencies** by default (that is what I wrote just above), it does not even read them **unless** you enable the `allowPeerDependenciesUpdate`. Since you never added the flag, your peer dependencies remained completely ignored and untouched. I updated the readme documentation with this missing information (you should read the docs again since I updated them). 

Lerna-Lite is a fork of Lerna and inherits its implementation, the original author completely ignored peer dependencies, which is why the `allowPeerDependenciesUpdate` option was added into Lerna-Lite to start reading and bumping peer dependencies, but again without this option enabled it will **never read or change peer dependencies** (and since it never reads or touch them, whatever text you put in these peer dependencies, like `workspace:*`, it will be published **as is**)

> 1. `lerna publish` does not have a such flag called `allowPeerDependenciesUpdate`,

`publish` automatically inherits all `version` options because Publish is using Version behind the scene.

> So, I do not accept your announcement that with the flag then peerDeps are good. I believed that these versions will still be released leaving the `workspace:*` in peerDeps as is.

Seriously I don't care that you don't accept this or not. This is an open source library with an MIT license which strictly says

> THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.

> * Are you suggesting I must add other things like `--graphType=allPlusPeerDependencies` to make it work

**No** you cannot do that, I already mentioned above that `allPlusPeerDependencies` is internal code and can **only** be changed via `allowPeerDependenciesUpdate`.

> * Shall this be a bug while `pnpm publish` support this?

again I don't care since this library is provided **as is**, like or not Lerna/Lerna-Lite completely ignores reading or touching peer dependencies and I don't see any good alternative to this problem. Unless you enable `allowPeerDependenciesUpdate` which you still haven't tried.

Seriously I don't understand why you never want to enable the flag and all you do so far is to complain. Why don't you start contributing or propose something that could help with the problem instead of always complaining? I assume English is not your first language but still, I am trying to stay friendly but it is really hard with your negative attitude.

I have been trying to tell you since the beginning to enable `allowPeerDependenciesUpdate` by adding it to your `lerna.json` configuration and so far I don't believe that you tried it even once yet? Did you? Please try to enable `allowPeerDependenciesUpdate` and if you still have problems then we can discuss further

**Comment by ghiscoding** [comment]
> With my test branch, you can see that I added the flag through the command line, as long as there is no special reminder in the docs (And also only one of these publish sources says inherited from `version` commans), Without your last reply. My next release will still probably be broken. I think the fault should be most on the documentation, the words are very misleading.

I looked at all your test commits and I did not see the `allowPeerDependenciesUpdate` flag anywhere in your code or in any of your tests since your lerna config has the following (`allowPeerDependenciesUpdate` is missing)

```json
{
  "npmClient": "pnpm",
  "version": "2.0.0-rc.36",
  "command": {
    "version": {
      "allowBranch": "lerna-test",
      "conventionalCommits": true,
      "exact": true,
      "message": "build: publish v%v",
      "syncWorkspaceLock": true
    }
  },
  "packages": ["plugins/*/*", "themes/*", "tools/*"],
  "ignoreChanges": ["**/tests/**", "**/*.md"]
}
```

I did not look too much at your shell command but from what I can see in your current script, the flag is not in your npm scripts either (so again `allowPeerDependenciesUpdate` is missing)
```sh
"release:publish": "lerna publish --dist-tag next from-package --yes && tsx scripts/syncNpmmirror.ts",
"release:version": "lerna version",
```
Since I don't see the flag anywhere in these 2, I imagine that you did not try the flag yet. 

About the documentation, those are all coming from the original Lerna project, if you think some section could be improved then please submit a PR. I don't understand what you're confused about from the screen shot above. Personally I only ever used the `from-package` and never tried anything else. 

I created Lerna-Lite fork when Lerna was no longer being maintained, my main goal was to update all its dependencies and make it more lightweight by making all commands as optional. I certainly don't know and never tried all of its options. I'm also just a simple Lerna user as well, I always use the same config for all projects that require Lerna and yes there are some options I don't quite understand and I don't always understand the original Lerna implementation. I'm just trying my best to support an existing library that was created by a smarter person than me. 

#### Different approach

Another approach to your original issue might be that when `allowPeerDependenciesUpdate` is missing or disabled, it would read the peer dependencies **only** to detect and replace any `workspace:` when found and not touch anything else. This mean that `workspace:` would be replaced properly and other peer deps would not be bumped or touched (unless the user enables `allowPeerDependenciesUpdate` which in that case would bump these peer deps as well). At least with this approach, it would not publish any `workspace:` by mistake. However, I'm not sure if this is doable and I would need to look at the code again and also for this approach to be accepted, I would have to make sure that it does not introduce breaking or behavior changes. So it is not a guarantee that will be doable but I can take a look at the code again

**Comment by ghiscoding** [comment]
Ok so I think that my new approach, as mentioned just above, is better and the correct approach to fix the problem. I tested it locally in Lerna-Lite with dry-run mode and below is the result when trying with and without the flag

#### input
you can see a few different types of `workspace:` are being tested
```ts
  "dependencies": {
    "@lerna-lite/cli": "workspace:*",
    "@lerna-lite/core": "workspace:^",
    "@lerna-lite/filter-packages": "workspace:~",
    "@lerna-lite/profiler": "workspace:*"
  },
  "peerDependencies": {
    "@lerna-lite/core": "workspace:*",
    "@lerna-lite/cli": "workspace:^",
    "@lerna-lite/profiler": "^3.0.0",
    "typescript": "^5.0.0"
  }
```

#### result without `allowPeerDependenciesUpdate` (missing or false)
you can see below that `cli` and `core` got bumped **but** `profiler` is untouched and remains exactly the same because the flag is disabled and only `workspace:` will be evaluated but anything else will be skipped.

```ts
"dependencies": {
    "@lerna-lite/cli": "3.5.1",
    "@lerna-lite/core": "^3.5.1",
    "@lerna-lite/filter-packages": "~3.5.1",
    "@lerna-lite/profiler": "3.5.1"
  },
  "peerDependencies": {
    "@lerna-lite/cli": "3.5.1",
    "@lerna-lite/core": "^3.5.1",
    "@lerna-lite/profiler": "^3.0.0",
    "typescript": "^5.0.0"
  },
```
#### result **with** `allowPeerDependenciesUpdate: true`
you can see below that this time all the peer deps got bumped including `@lerna-lite/profiler` package which got bumped this time
```ts
  "dependencies": {
    "@lerna-lite/cli": "3.5.1",
    "@lerna-lite/core": "^3.5.1",
    "@lerna-lite/filter-packages": "~3.5.1",
    "@lerna-lite/profiler": "3.5.1"
  },
  "peerDependencies": {
    "@lerna-lite/cli": "3.5.1",
    "@lerna-lite/core": "^3.5.1",
    "@lerna-lite/profiler": "3.5.1",
    "typescript": "^5.0.0"
  },
```

I think this is the correct approach this time, I do have more work to make sure that all unit tests are passing.

**Comment by Mister-Hope** [comment]
> I looked at all your test commits and I did not see the allowPeerDependenciesUpdate flag anywhere in your code or in any of your tests since your lerna config has the following (allowPeerDependenciesUpdate is missing)

Hi, this is want I am doing https://github.com/vuepress/ecosystem/commit/011026032656cf6233fc00a8f8314de984ea6d1d so that's why I want to improve current docs.
